### PR TITLE
[RNMobile] Add "Set as Featured Image" Button to Image Block (Android Only)

### PR DESCRIPTION
**Pull request by SiobhyB**
Edit: This PR was closed as it was separated into the following individual PRs:

* **Add Featured Banner to Image Block (Android Only):** #30806
* **Add "Set as Featured Image" Button to Image Block (Android Only):** #31705

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java
Comment (jd-alexander) [inline packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java]: Hi @SiobhyB we could also add this override to the [MainApplication](https://github.com/WordPress/gutenberg/blob/702c76e9e9b0de4f96b324f017716afa333f421a/packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java) class of the demo project if it's missing. Let me know if you get the project to build in Android Studio. :)

**Comment by jd-alexander** [comment]
@SiobhyB the bridge code is looking good so far 😄 . I am seeing some build errors from the Android side of things and it seems related to the click listeners, so let me know if you are able to resolve them. Thanks 👍

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java
Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java]: Added the missing function to MainApplication.java. :)

**Comment by paaljoachim** [comment]
Hi

Could you add a screenshot?
I wonder if the method could also be transferred over to the desktop Image block.
Thanks.

**Comment by SiobhyB** [comment]
👋 @paaljoachim, sure, the mockup of what this will likely look like is as follows: 

58123648-ef9dfa80-7bd1-11e9-9095-3767b85f69e4 (1)

I'll defer to @iamthomasbishop for any extra details around the design and some extra discussion can be found in [the mobile repo](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1011). Hope that helps!

**Comment by SiobhyB** [comment]
@paaljoachim, hi! :) I just wanted to follow up again to check whether there are any upcoming plans to implement this on the web? We don't currently have an ETA on the mobile side but would be responsive and adjust our plans to work with the web, if there are firm plans in place there.

**Comment by paaljoachim** [comment]
Hi @SiobhyB 

This is the issue
**Add option to set as feature image to image / cover block**
https://github.com/WordPress/gutenberg/issues/13795

I have earlier added a checkbox below the [Featured image placeholder box mockup](https://github.com/WordPress/gutenberg/issues/13795#issuecomment-636787215). Some what similar to the first proposal mockup you share. I did not extend it beyond that point. What about creating another proposal set with focus on web and share them in the above issue? We can then ping @shaunandrews Shaun to give some feedback. This way we will have a have a similar method to automatically add a featured image for web and Mobile views.
Thanks!

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java
Comment (hypest) [inline packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java]: 👋 @SiobhyB , not sure, is the removal of the `onFocalPointPickerTooltipListener` parameter intentional? It seems to introduce a compile error since `onFocalPointPickerTooltipListener` is then not found in line 601. 🤔

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java]: @hypest, I _believe_ that came about by mistake following a previous merge conflict. I had fixed the issue in my local environment but hadn't committed, done that now! Thanks for the ping. :)

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (hypest) [inline packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: A bit of a nit but, looks like this is a "change happened" event (judging from the "Change" part of the string literal), but the `EVENT_FEATURED_IMAGE_ID` name is a bit more vague. What do you think about renaming to `EVENT_FEATURED_IMAGE_ID_CHANGED` to be more precise about what this event is denoting?

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (hypest) [inline packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Similar to my naming comment above, what do you think about renaming to `featuredImageIdChanged()` or better:`onFeaturedImageIdChange` to make it read like an event handler? The original `featuredImageIdNotifier` makes me wonder what gets notified... like, is it the user that receives a notification when this function is called?

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Reading more code where this function is used, I see that it's also called when the JS side [just asks for an update of the featured image id](https://github.com/wordpress-mobile/WordPress-Android/blob/2700d6c49232e6866abdc796cf7cefb5f8c26dea/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L296). So, the function's purpose is not only to notify the JS side that the id has changed, which means that the event name ([defined here](https://github.com/WordPress/gutenberg/pull/28854/files#diff-bd64b4a30ececb510793aeabfe0b68175eaa825a9f121d8dd953e688a355c01fR46)) is not actually accurate 🤔.

So, if I understand things properly, the event is all in all just trying to pass the the media id of the current featured image, right, whether it was recently updated or not? If so, what do you think about naming the event with `EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent"`?

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: I think I have a new understanding of what this event is about, please see my newer proposal for the event name in [the comment bellow](https://github.com/WordPress/gutenberg/pull/28854/files#r598733588), sorry for the noise!

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java
Comment (hypest) [inline packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java]: Hey Siobhan, aha, can you elaborate on why this new interface is commented out and the new featured image related interface functions are placed in `OnMediaLibraryButtonListener`? I'm not sure if `OnMediaLibraryButtonListener` is actually related 🤔, I'd expect this new/separate interface definition to be the desired implementation.

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: > The original featuredImageIdNotifier makes me wonder what gets notified... like, is it the user that receives a notification when this function is called?

This is useful feedback, thank you @hypest! I looked through other similar functions and renamed `featuredImageIdNotifier` to `onRequestFeaturedImageId` in order to hopefully be clearer of the function's purpose. Let me know if you think this is still unclear!

> So, if I understand things properly, the event is all in all just trying to pass the the media id of the current featured image, right, whether it was recently updated or not? If so, what do you think about naming the event with EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent"?

That makes sense! The purpose changed as the code evolved (it was originally only passing the id when the image changed) which is why I think the naming became a bit messy/confusing. This is a good lesson to keep on top and iterating on names. Thank you! I've updated this to "featuredImageIdCurrent" now.

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Great, I've responded to the other comment. :)

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java
Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/WPAndroidGlueCode.java]: @hypest, this one was under the "I'll get to that later!" list in my mind that I hadn't got round to yet. I believe I was just exploring a few things to see if they worked when I placed the functions under `OnMediaLibraryButtonListener` and hadn't got round to implementing `OnSetFeaturedImageListener` yet. I've done that now. :) Thanks for the push!

File: packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java
Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: > I looked through other similar functions and renamed featuredImageIdNotifier to onRequestFeaturedImageId in order to hopefully be clearer of the function's purpose. Let me know if you think this is still unclear!

Aha.

Still trying to grok what the code is trying to do, please bear with me :). There are two rather different code paths that use this function:
1. The JS side asks the native side to provide the featured image id ([getter named function here](https://github.com/WordPress/gutenberg/blob/46ba7e8d770cc0cc1a93664047fdcaaa12f0f0f0/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L237-L239)). The native side then [calls `onRequestFeaturedImageId` only if the media id provided matches the featured image id on file](https://github.com/wordpress-mobile/WordPress-Android/blob/622dfad0700d288349fd344783b3bbf24e242cdc/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L349-L353). By the way, looks like it's not a getter after all since the response to the original `getFeaturedImageId` is conditional. Looks to me more like a "check for something", not a plain getter. If so, would it make sense to rename getFeaturedImageId -> checkIfFeaturedImage?
2. The native side [sends the id when the featured image is set](https://github.com/wordpress-mobile/WordPress-Android/blob/622dfad0700d288349fd344783b3bbf24e242cdc/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2554).

From flow No2 above, and the function implementation itself ([native side](https://github.com/WordPress/gutenberg/blob/d8c0824c087e5fc76e955123665bc9f2afc893e8/packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java#L214-L216), https://github.com/WordPress/gutenberg/blob/75de59fd6cacd8faeb787e20c8bc218c24992f46/packages/block-library/src/image/edit.native.js#L305-L319) I understand that onRequestFeaturedImageId just sends the current featured image id to the JS side so, may I propose renaming it to `sendToJSFeaturedImageId`?

Then, flow No1 shows to me we need to better name and document why `onGetFeaturedImageId` doesn't always call sendToJSFeaturedImageId.
Does this all make sense? Sorry if I'm a bit confused on this one.

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: @hypest, thank you again for your help here!  

> By the way, looks like it's not a getter after all since the response to the original getFeaturedImageId is conditional. Looks to me more like a "check for something", not a plain getter. If so, would it make sense to rename getFeaturedImageId -> checkIfFeaturedImage?
>
> Then, flow No1 shows to me we need to better name and document why onGetFeaturedImageId doesn't always call sendToJSFeaturedImageId.

It's right that the `getFeaturedImageId` function is only checking whether an image is featured, I've changed its name to `checkIfFeaturedImage` for the time being.

To further clarify: The `checkIfFeaturedImage` function is called within `componentWillUnmount()` on the JS side. It's currently responsible for flagging which image is featured when the editor is first loaded.

The background for this is that I was unable to get `getEditedPostAttribute` to work to get the post's initial featured image (discussed on project thread and in project's Slack channel). 

I've spent a bit of time with this but am not sure if there's a more elegant way to achieve the same effect, but it's entirely likely I've been looking at the code too long! I'd be interested to hear whether you think this approach makes sense, with those extra details, or if I should work on this a bit more.

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: > To further clarify: The checkIfFeaturedImage function is called within componentWillUnmount() on the JS side. It's currently responsible for flagging which image is featured when the editor is first loaded.

Ah, I see 👍, I will check this out now that I read the logic better. Just to make sure, do you perhaps meant to say `componentDidMount` since that's the one called on editor start?

> I've spent a bit of time with this but am not sure if there's a more elegant way to achieve the same effect, but it's entirely likely I've been looking at the code too long! I'd be interested to hear whether you think this approach makes sense, with those extra details, or if I should work on this a bit more.

Will have a look from than angle and let you know! Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks! ([Commented on the WPAndroid PR as well](https://github.com/wordpress-mobile/WordPress-Android/pull/14052#issuecomment-810152596))

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: > Just to make sure, do you perhaps meant to say componentDidMount since that's the one called on editor start?

Ah, yes, careless copy/paste, sorry! 🤦‍♀️ I meant `componentDidMount`.

> Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks!

Updated! :)

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: > Will have a look from than angle and let you know! Can you update the WPAndroid side as well as I think it doesn't compile currently due to bridge changes, thanks! (Commented on the WPAndroid PR as well)

Just an update. I had a short look and I feel like triggering a bridge method to check each image in the post is too much and it might not scale well in image heavy posts. Feels to me that we should pass the initial value of the featured image id during editor boot, find a way to keep that in the global(?) state and check against that for every image block.

To that end, I'm having another look to the original idea around `getEditedPostAttribute` to see if we could hook at that place, to put the id in the global post object, for example. I'm not familiar with that part of the store though so, will continue exploring into it.

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Did a quick dive and I think that adding a `featured_media: props.featuredMedia` in [the post object we prepare when starting the editor](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/editor.native.js#L122-L136), accompanied by passing that `featuredMedia` id [down the bridge](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt), is then making `featured_media` available to the `getEditedPostAttribute` Store function. Not sure if that's something you already had tried recently @SiobhyB , let me know if it helps.

That's something to help with the bootstrapping of the editor with the featured image id becoming available then, but will need some more treatment to enable _updating_ the id when the featured image gets set after the editor has already started.

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: @hypest, thank you for this! 🙇‍♀️ I'm working on getting this to work as expected now. I've committed my latest work in progress, which is not yet working, here: https://github.com/WordPress/gutenberg/pull/28854/commits/0cdeadb1ddf8cbc4c491e2eb21b42eda683ac247

I also made the assumption that this will require some changes to the Android side, and have made some initial changes here: https://github.com/wordpress-mobile/WordPress-Android/pull/14052/commits/0ec24df06d5ff419e0f7a21849c1e53afdbf70a8

Does this seem like I'm on the right path? I'm happy to continue troubleshooting and trying to figure this out, but just wanted to make sure I wasn't going off in the wrong direction.

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Oh, glad you liked the proposal and started going down that route!

Looks like a promising path so far so yeah, please feel free to go ahead. A disclaimer here though: I haven't really worked on getting a real feel of how the _updating_ part will work this time so, please keep that in mind in case this new route proves impractical too after all. Thanks!

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: I've been able to get this working to grab the post's initial featured image ID now and have made use of `setAttributes` to get this working in ComponentDidUpdate [here](https://github.com/WordPress/gutenberg/pull/28854/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R197). How does this seem? Looking forward to any feedback or ideas on this one. :)

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: > How does this seem?

Yeah, I think we're on a good route here. We chatted a bit over Zoom, to share my thoughts on how the ImageEdit component just needs to listen to the store updates and compare the featured image id it gets from there with the id it keeps in the block attributes, and use those to compute a "isFeaturedImage" boolean inside the render() function.

Then, it becomes a matter of figuring out how to also do the update of the id in the store, for which we can use the [editEntityRecord()](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/actions.js#L248) action creator (acquiring it via the `withDispach`. But, that part (to update the featured image id value when the image changes) can be done in a separate PR I think.

Let me know what you think!

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Thank you again for this! I've spent a bit of time today trying to process this and have a few questions:

* What are the overall benefits to moving the `setFeaturedImageState()` functionality inside the `render()` function vs. the current approach? 
* `setFeaturedImageState()` is currently called in `componentDidMount()` for when the editor first loads and then `componentDidUpdate()` for when an image is replaced directly within the image block (using the Edit icon to the upper right when the block is selected). Will moving the functionality for setting state from `setFeaturedImageState()` to `render()` resolve the need to do that? Is that one of the benefits?
* It seems like I'll need to figure out how to use `editEntityRecord()` to grab the post's current featured image ID _before_ I can move the functionality for setting state from  `setFeaturedImageState()`. Is that right?

Sorry for my fuzziness on this one, just want to make sure I'm clear before moving ahead!

Comment (SiobhyB) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: It seems all I had to do was ask those questions and then play around a little more to figure out the answers. :) I've been able to convert the functionality previously included in `setFeaturedImageState` to an `isFeaturedImage` boolean. I can see now that the answer to my second question is "yes" and the code is a lot cleaner as a result. 🎉 I've committed those changes.

Comment (hypest) [reply packages/react-native-bridge/android/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java]: Nice to see the simplification of `isFeaturedImage`!

> setting an isFeaturedImage state

Just a quick comment on the terminology used in the [commit message](https://github.com/WordPress/gutenberg/pull/28854/commits/39303e4a13dce66bfeed96060e452af72e65fb53): now that `isFeaturedImage` is just a helper local const in `render()`, referring to it as "state" can be misleading, mainly because within React, "state" has a special and specific [meaning](https://reactjs.org/docs/state-and-lifecycle.html). I'd recommend referring to it as "local variable/const" to avoid confusion with React's use.

**Review by bakoorahnin — APPROVED** [review]

**Comment by SiobhyB** [comment]
@hypest, @iamthomasbishop, the current PR includes the **Featured** banner for Android users. It then also adds the **set as featured** functionality to the image block's settings for Android users in development mode. The screenshots in this issue's main description are up-to-date for Android. If this looks good to merge to you, I'll then work on the following in separate PRs:

* Introduce the **Featured** banner for iOS users.
* Introduce the **set as featured** functionality to the image block's settings for iOS users.
* Improve the way the featured image ID is managed on the JS side.
* Remove the dev-only flag.

Looking forward to hearing what you think!

**Comment by SiobhyB** [comment]
Following a recent discussion, I've now split the **Featured** banner part of this off into its own PR here: https://github.com/WordPress/gutenberg/pull/30806

**Comment by SiobhyB** [comment]
@jd-alexander, this chain of PRs is ready for review, whenever you have the chance to take a look. Thank you in advance! 🙇‍♀️

**Comment by jd-alexander** [comment]
> @jd-alexander, this chain of PRs is ready for review, whenever you have the chance to take a look. Thank you in advance! 🙇‍♀️

Thanks! I will have a look.

File: packages/block-library/src/image/edit.native.js
Comment (hypest) [inline packages/block-library/src/image/edit.native.js:505]: 👋 Siobhan! I see the [same line in trunk already](https://github.com/WordPress/gutenberg/blob/f14962f283f63039be1e790f552489a1705d2928/packages/block-library/src/image/edit.native.js#L451) so, I assume the PR needs some updating from trunk? While at it, since by now you have write access to the Gutenberg repo, it makes sense to re-open this PR as one directly on the repo instead of the fork, that will speed up the review process, wdyt?

File: packages/block-library/src/image/edit.native.js
Comment (hypest) [inline packages/block-library/src/image/edit.native.js:697]: Ah, [like recently](https://github.com/WordPress/gutenberg/pull/30806#discussion_r613207635), let's not use store name literals directly, thanks!

Comment (SiobhyB) [reply packages/block-library/src/image/edit.native.js:697]: Ah, thank you! I'll refactor this in the new PR I'm creating for the issue. :)

File: packages/block-library/src/image/edit.native.js
Comment (SiobhyB) [reply packages/block-library/src/image/edit.native.js:505]: Thank you for pointing that out, it looks like the line got moved about at some point in this PR, I'll rework that.

> it makes sense to re-open this PR as one directly on the repo instead of the fork, that will speed up the review process, wdyt?

This makes sense to me and will likely simplify the process of cleaning up the commit history! I'll get about doing that and ping you when the new PR is ready for review. :)

**Comment by SiobhyB** [comment]
Closing this PR, as it was separated into the following:

* **Add Featured Banner to Image Block (Android Only):** https://github.com/WordPress/gutenberg/pull/30806
* **Add "Set as Featured Image" Button to Image Block (Android Only):** https://github.com/WordPress/gutenberg/pull/31705

**Pull request closed without merging**
